core/finality-grandpa/src/authorities.rs

```diff
@@ -115,138 +122,131 @@ impl<H, N> AuthoritySet<H, N> {
 impl<H: Eq, N> AuthoritySet<H, N>
 where
 	N: Add<Output=N> + Ord + Clone + Debug,
-	H: Debug
+	H: Clone + Debug
 {
-	/// Note an upcoming pending transition. This makes sure that there isn't
-	/// already any pending change for the same chain. Multiple pending changes
-	/// are allowed but they must be signalled in different forks. The closure
-	/// should return an error if the pending change block is equal to or a
-	/// descendent of the given block.
-	pub(crate) fn add_pending_change<F, E: Debug>(
+	/// Note an upcoming pending transition. Multiple pending changes on the
+	/// same branch can be added as long as they don't overlap. This method
+	/// assumes that changes on the same branch will be added in-order. The
+	/// given function `is_descendent_of` should return `true` if the second
+	/// hash (target) is a descendent of the first hash (base).
+	pub(crate) fn add_pending_change<F, E>(
```

Comment:
I wonder if as a safety fallback we should cap the number of pending changes in a given fork. I need to think how this will interact with the offline fallback as well.